File: Patterns/Registration/Python/test_RegistrationDemo.py
from TestPackage.Base.ComponentBuilder import ComponentBuilder
from TestPackage.Base.ComponentRegistration import ComponentRegistration
import pytest

def test_registration_pattern():
    # Collection of Already Registered components
    expected = {
        "SubModule_01_01": "SubModule_01_01 Required",
        "SubModule_01_02": "SubModule_01_02 Required",
        "SubModule_01_02_extra": "SubModule_01_02_extra Required",
        "SubSubModule_01_02_01": "SubSubModule_01_02_01 Required",
        "SubModule_02": "SubModule_02 Required"
    }

    # Check that each invocation of builder gives expected result
    for key, value in expected.items():
        comp = ComponentBuilder.build(key)
        assert comp.required() == value

    # Try to use class not registered to builder
    with pytest.raises(Exception):
        ComponentBuilder.build('Decoy')

def test_register_class():
    from TestPackage.Base.ComponentBase import ComponentBase

    # Register New Class which inherits from Base
    @ComponentRegistration.register('MyClass')
    class MyClass(ComponentBase):
        def required(self):
            return "MyClass Required"

    # Check that new class is available to builder
    testComp = ComponentBuilder.build('MyClass')
    assert testComp.required() == "MyClass Required"

    # Registering a class which doesn't inherit from Base should fail
    with pytest.raises(Exception):
        @ComponentRegistration.register('MyClass2')
        class MyClass2():
            pass

if __name__ == "__main__":
    test_registration_pattern()
    test_register_class()

File: Patterns/Registration/Cpp/test_RegistrationDemo.cpp
#include "gtest/gtest.h"
#include "ComponentBuilder.h"
#include "ComponentRegistration.h"

TEST(RegistrationDemo, UseRegistration){
    // Collection of Already Registered components
    // std::map<std::string, std::string> expected = {
    //     {"SubModule_01_01", "SubModule_01_01 Required"},
    //     {"SubModule_01_02", "SubModule_01_02 Required"},
    //     {"SubModule_01_02_extra", "SubModule_01_02_extra Required"},
    //     {"SubModule_01_02_01", "SubModule_01_02_01 Required"},
    //     {"SubModule_02", "SubModule_02 Required"}
    // };
    std::map<std::string, std::string> expected = {
        {"SubModule_01_01", "SubModule_01_01 Required"}
    };

    // Check that each invocation of builder gives the expected result
    for (const auto item : expected){
        auto component = ComponentBuilder::build(item.first);
        EXPECT_EQ(component->required(), item.second);
    }

    // Try to use class not registered to builder
    ASSERT_THROW(ComponentBuilder::build("Decoy"), std::exception);
}

TEST(RegistrationDemo, RegisterClass){

    // Register new class which inherits from Base
    class MyClass : public ComponentBase {
        public:
            std::string required() {return "MyClass Required";}
    };
    singleton::registry().registerComponent("MyClass", [](){return std::make_shared<MyClass>();});

    // Check that new class is available to builder
    auto testComp = ComponentBuilder::build("MyClass");
    EXPECT_STREQ(testComp->required().c_str(), "MyClass Required");
}